Add sizes to Index to use in Indexer removing Deprecated annotations

matteodg opened the pull request:
Getting back `sizes()` in `Indexer` and in `Index` as the maximum values of the indices.

matteodg commented on src/main/java/org/bytedeco/javacpp/indexer/HyperslabIndex.java:
```diff
@@ -50,13 +50,19 @@
+
+        this.resultingSizes = new long[counts.length];
+        for (int i = 0; i < counts.length; i++) {
+            resultingSizes[i] = counts[i] * blocks[i];
```
Just to not cause confusion with the `sizes` of `StrideIndex` as they are not the same as the resulting sizes of the hyperslab selection.

saudet replied:
Ok, but there are not other `sizes` in `HyperslabIndex`, so I don't think they can be confused.

matteodg replied:
The sizes in the `Index` are the ones used to reshape it, whereas the sizes in the `Hyperslab` are actually the result of a sub-selection from the underlying reshaped flat array/buffer/pointer.

Maybe the reshaping and sub-selection should actually be clearly separated.

saudet replied:
Right, but the `sizes` that the user provides are used exactly no where in `HyperslabIndex`, so why do we care about them?

saudet replied:
Oh, I see, they get used to compute the `strides`. Shouldn't users be allowed to set those then?

matteodg replied:
Something like this? I added an `HyperslabIndex` constructor to allow to set the strides for reshaping.

saudet replied:
Something like this yes, but let's make the tests not fail.

saudet commented:
I've made a few changes such that `HyperslabIndex` actually just updates the `sizes`. I think that's OK since they do not get used by `Indexer`. In other words, it's only useful to users. What do you think?

matteodg commented:
I just rebased on latest master and added generics to `reindex()`.

> I've made a few changes such that HyperslabIndex actually just updates the sizes. I think that's OK since they do not get used by Indexer. In other words, it's only useful to users. What do you think?

It is OK, I do not particularly like using `sizes` for two different purposes (reshaping and sub-selecting). And I guess you kept the `sizes` field and set to `index.sizes()` here:
```
    protected Indexer(Index index) {
        this.index = index;
        this.sizes = index.sizes();
        if (index instanceof StrideIndex) {
            this.strides = ((StrideIndex)index).strides();
        }
    }
```
so whoever extended `Indexer` can rely on the `sizes` protected field, that's fine with me at the end.

saudet commented:
Yeah, it's not too clear what would be best here, but unless problems arise using the `sizes` field as output that way, I'd like to see how this goes. The thing is, we're not actually reshaping or selecting anything. When calling say `reindex()`, it's basically just a new view on the same data. We can think of it either as reshaping or (sub)selecting I suppose, but we're not actually differentiating between them. If we were to think of these 3 concepts separately as with NumPy, it's not always possible to use them together. For example, reshaping a slice doesn't always results in a view: Sometimes we need a copy. I'd rather just think of everything as views, at least for a simple API like Indexer. :)

BTW, we don't need the sizes in `Indexer` anymore because we can call `index.sizes()`. It's there just in case, but deprecated... Sounds OK?

matteodg commented:
> If we were to think of these 3 concepts separately as with NumPy, it's not always possible to use them together. For example, reshaping a slice doesn't always results in a view: Sometimes we need a copy. I'd rather just think of everything as views, at least for a simple API like Indexer. :)

If I understand correctly with Foreign Memory API in JDK15 we will be able to create views in all cases (using a `VarHandle` to a Java array as well as to `Buffer`, `Pointer` or anything), that would be great!

One side question: are there somewhere the artifact of the current master branch? I cannot find them in Travis. I'm having troubles compiling for `linux-x86_64`: I tried to use the commands I found in `.travis.yml` but something is off (I still have to dig into it).

saudet commented:
> If I understand correctly with Foreign Memory API in JDK15 we will be able to create views in all cases (using a `VarHandle` to a Java array as well as to `Buffer`, `Pointer` or anything), that would be great!

Not sure about whether or not it's going to be in JDK 15, but it will become possible at some point, yes. It still won't work with other implementations of Java such as Android though...
 
> One side question: are there somewhere the artifact of the current master branch? I cannot find them in Travis. I'm having troubles compiling for `linux-x86_64`: I tried to use the commands I found in `.travis.yml` but something is off (I still have to dig into it).

Yes, snapshot builds are available: http://bytedeco.org/builds/

matteodg commented:
Thanks for the snapshot builds link, it works like a charm
